src/blockchain/rpc.rs

```diff
@@ -495,6 +285,507 @@ fn list_wallet_dir(client: &Client) -> Result<Vec<String>, Error> {
+                    txid: tx_res.info.txid,
+                    ..Default::default()
+                }
+            });
+
+            // update raw tx (if needed)
+            let raw_tx =
+                &*match &mut db_tx.transaction {
+                    Some(raw_tx) => raw_tx,
+                    db_tx_opt => {
+                        updated = true;
+                        db_tx_opt.insert(client.get_raw_transaction(
+                            &tx_res.info.txid,
+                            tx_res.info.blockhash.as_ref(),
+                        )?)
+                    }
+                };
+
+            // update fee (if needed)
+            if let (None, Some(new_fee)) = (db_tx.fee, tx_res.detail.fee) {
+                updated = true;
+                db_tx.fee = Some(new_fee.as_sat().unsigned_abs());
+            }
+
+            // update confirmation time (if needed)
+            let conf_time = BlockTime::new(tx_res.info.blockheight, tx_res.info.blocktime);
+            if db_tx.confirmation_time != conf_time {
+                updated = true;
+                db_tx.confirmation_time = conf_time;
+            }
+
+            // update received (if needed)
+            let received = Self::received_from_raw_tx(self.db, raw_tx)?;
+            if db_tx.received != received {
+                updated = true;
+                db_tx.received = received;
+            }
+
+            // check if tx has an immature coinbase output (add to updated UTXOs)
+            // this is required because `listunspent` does not include immature coinbase outputs
+            if tx_res.detail.category == GetTransactionResultDetailCategory::Immature {
+                let txout = raw_tx
+                    .output
+                    .get(tx_res.detail.vout as usize)
+                    .cloned()
+                    .ok_or_else(|| {
+                        Error::Generic(format!(
+                            "Core RPC returned detail with invalid vout '{}' for tx '{}'",
+                            tx_res.detail.vout, tx_res.info.txid,
+                        ))
+                    })?;
+
+                if let Some((keychain, index)) =
+                    self.db.get_path_from_script_pubkey(&txout.script_pubkey)?
+                {
+                    let utxo = LocalUtxo {
+                        outpoint: OutPoint::new(tx_res.info.txid, tx_res.detail.vout),
+                        txout,
+                        keychain,
+                        is_spent: false,
+                    };
+                    self.updated_utxos.insert(utxo);
+                    self.update_last_index(keychain, index);
+                }
+            }
+
+            // update tx deltas
+            self.retained_txs.insert(tx_res.info.txid);
+            if updated {
+                self.updated_txs.insert(tx_res.info.txid);
+            }
+        }
+
+        // update tx sent fields from tx inputs
+        self.txs
```

Comment:
I have a couple of suggestions here to improve this a bit:

1. instead of cloning the whole thing you can call `values()` first, so at least you avoid cloning the keys
2. the check that looks for the tx in `retained_txs` could be moved into a `filter()` method, which maybe makes this a bit more readable. Also this check could be moved before the `clone()` to again avoid cloning things you don't need

Reply:
@afilini Thank you for the suggestions!

However, using `.values().cloned()` would not work, as calling `.cloned()` only clones one element at a time (and we have a mutable reference of the whole `self`).

In my original commit, I had all filters in an `.filter_map` and had two loop (one for marking changes and one for applying changes) so we avoid the mutable-reference problems.

I think it is best to kinda go back to my original solution, but improve commenting and split "filter logic" into multiple `.filter..()` calls.

This is what I have now:

```rust
        // obtain vector of tx-sent-value changes
        let sent_updates = self
            .txs
            .values()
            // only bother to update txs that are retained
            .filter(|db_tx| self.retained_txs.contains(&db_tx.txid))
            // only bother to update txs where the raw tx is accessable
            .filter_map(|db_tx| (db_tx.transaction.as_ref().map(|tx| (tx, db_tx.sent))))
            // recalcuate sent value, only update txs in which sent value is changed
            .filter_map(|(raw_tx, old_sent)| {
                self.sent_from_raw_tx(raw_tx)
                    .map(|sent| {
                        if sent != old_sent {
                            Some((raw_tx.txid(), sent))
                        } else {
                            None
                        }
                    })
                    .transpose()
            })
            .collect::<Result<Vec<_>, _>>()?;

        // record send updates
        sent_updates.iter().for_each(|&(txid, sent)| {
            // apply sent field changes
            self.txs.entry(txid).and_modify(|db_tx| db_tx.sent = sent);
            // mark tx as modified
            self.updated_txs.insert(txid);
        });
```

This was my original:

https://github.com/bitcoindevkit/bdk/blob/9d787bf122edf9e5aa6fb43f2315860c1bfb732a/src/blockchain/rpc.rs#L454-L476

Reply:
I just had a thought. If we can guarantee that txs returned from `listtransactions` RPC call are in chronological order (which I assume it would be), we can have the `TransactionDetails::sent` recalculations in the same loop as the recalculations of all the other fields.

The only benefit of this is for a slight performance gain, so maybe no point to bother? (I would assume that reducing IO would contribute the most to performance)

Reply:
Yeah I remember this now, I think I was the one who suggested merging the two loops :sweat_smile: 

In the end I like what you came up with!